components/timezone/src/ids.rs

```diff
@@ -0,0 +1,371 @@
+/// assert_eq!(
+///     mapper.iana_to_bcp47("australia/melbourne"),
+///     Some("aumel".parse().unwrap())
+/// );
+///
+/// // The IANA zone "Australia/Victoria" is an alias:
+/// assert_eq!(
+///     mapper.iana_to_bcp47("Australia/Victoria"),
+///     Some("aumel".parse().unwrap())
+/// );
+///
+/// // We can recover the canonical name from the mapper:
+/// assert_eq!(
+///     mapper
+///         .canonicalize_iana("Australia/Victoria")
+///         .unwrap()
+///         .string,
+///     "Australia/Melbourne"
+/// );
+/// ```
+#[derive(Debug, Clone)]
+pub struct TimeZoneIdMapper {
+    data: DataPayload<IanaToBcp47MapV2Marker>,
+}
+
+impl TimeZoneIdMapper {
+    /// Creates a new [`TimeZoneIdMapper`] using compiled data.
+    ///
+    /// See [`TimeZoneIdMapper`] for an example.
+    ///
+    /// ✨ *Enabled with the `compiled_data` Cargo feature.*
+    ///
+    /// [📚 Help choosing a constructor](icu_provider::constructors)
+    #[cfg(feature = "compiled_data")]
+    pub fn new() -> Self {
+        Self {
+            data: DataPayload::from_static_ref(
+                crate::provider::Baked::SINGLETON_TIME_ZONE_IANA_TO_BCP47_V2,
+            ),
+        }
+    }
+
+    icu_provider::gen_any_buffer_data_constructors!(locale: skip, options: skip, error: DataError,
+        #[cfg(skip)]
+        functions: [
+            new,
+            try_new_with_any_provider,
+            try_new_with_buffer_provider,
+            try_new_unstable,
+            Self,
+        ]
+    );
+
+    #[doc = icu_provider::gen_any_buffer_unstable_docs!(UNSTABLE, Self::new)]
+    pub fn try_new_unstable<P>(provider: &P) -> Result<Self, DataError>
+    where
+        P: DataProvider<IanaToBcp47MapV2Marker> + ?Sized,
+    {
+        let data = provider.load(Default::default())?.take_payload()?;
+        Ok(Self { data })
+    }
+
+    /// Returns a borrowed version of the mapper that can be queried.
+    ///
+    /// This avoids a small potential cost of reading the data pointer.
+    pub fn as_borrowed(&self) -> TimeZoneIdMapperBorrowed {
+        TimeZoneIdMapperBorrowed {
+            data: self.data.get(),
+        }
+    }
+}
+
+/// A borrowed wrapper around the time zone ID mapper, returned by
+/// [`TimeZoneIdMapper::as_borrowed()`]. More efficient to query.
+#[derive(Debug, Copy, Clone)]
+pub struct TimeZoneIdMapperBorrowed<'a> {
+    data: &'a IanaToBcp47MapV2<'a>,
+}
+
+impl<'a> TimeZoneIdMapperBorrowed<'a> {
+    /// Gets the BCP-47 time zone ID from an IANA time zone ID
+    /// with a case-insensitive lookup.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_timezone::TimeZoneBcp47Id;
+    /// use icu_timezone::TimeZoneIdMapper;
+    /// use tinystr::tinystr;
+    ///
+    /// let mapper = TimeZoneIdMapper::new();
+    /// let mapper = mapper.as_borrowed();
+    ///
+    /// assert_eq!(
+    ///     mapper.iana_to_bcp47("AMERICA/chicago"),
+    ///     Some(TimeZoneBcp47Id(tinystr!(8, "uschi")))
+    /// );
```

Comment:
suggestion: add an example with a non-canonical iana that fails

Reply:
Fixed